napari/_qt/widgets/qt_dims_sorter.py

```diff
@@ -46,31 +46,33 @@ class QtDimsSorter(QWidget):
     Modified from:
     https://github.com/jni/zarpaint/blob/main/zarpaint/_dims_chooser.py
     """
-
     def __init__(self, viewer: 'Viewer', parent=None) -> None:
         super().__init__(parent=parent)
         dims = viewer.dims
-        root = AxisList.from_dims(dims)
-        root.events.reordered.connect(
-            lambda event: set_dims_order(dims, event.value)
-        )
-        dims.events.order.connect(
-            lambda event: move_indices(root, event.value)
-        )
-        view = QtListView(root)
-        view.setSizeAdjustPolicy(QtListView.AdjustToContents)
+        self.axis_list = AxisList.from_dims(dims)
 
-        self.axes_list = root
+        view = QtListView(self.axis_list)
+        view.setSizeAdjustPolicy(QtListView.AdjustToContents)
 
         layout = QGridLayout()
         self.setLayout(layout)
 
         widget_tooltip = QtToolTipLabel(self)
         widget_tooltip.setObjectName('help_label')
-        widget_tooltip.setToolTip(trans._('Drag dimensions to reorder.'))
+        widget_tooltip.setToolTip(trans._('Drag dimensions to reorder, uncheck to lock dimension in place.'))
 
         widget_title = QLabel(trans._('Dims. Ordering'), self)
 
         self.layout().addWidget(widget_title, 0, 0)
         self.layout().addWidget(widget_tooltip, 0, 1)
         self.layout().addWidget(view, 1, 0)
+
+        # connect axes_list and dims
+        self.axis_list.events.reordered.connect(
+            lambda event: set_dims_order(dims, event.value), 
+            until=self.parent().finished
+        )
+        dims.events.order.connect(
```

Comment:
Great catch here! Just to confirm, the reason for the problem is because of multiple connections to `dims.events.order` because there's only one instance of `Dims` per `Viewer`?

I don't entirely understand your solution yet, but it seems promising. I just became aware of `until`, but that seems like a reasonable approach. What is `finished`? I looked through the Qt docs, but couldn't find it. Also, is it possible for `parent` to be `None` here?

Reply:
With every opening of the popup more and more event connections pileup. Thus, we need to terminate them the moment the popup is closed. The popup are implemented as a subclass of QDialog (Qt class) and QDialog emits an event the moment it is closed. This event contains the return value of the QDialog, I don't use this but just catch the event to use it for the event terminations.

Reply:
To be able to turn parent to None, we would need to add an additional parameter to QtDimsSorter. __init__ to be able to use the signal of the popup. In the previous implementation QtDimsSorter had the viewer as parent, but that is not necessary as the popup has the viewer as parent.

Reply:
Ah, that helps clarify. So given the surrounding code, we know that `self.parent()` will return an instance of `QDialog`? It would be super useful to clarify that with a type annotation (or maybe better with `typing.cast`) because in general (i.e. without prior knowledge) `QWidget.parent()` returns an `Optional[QObject]`.

Again, no need to do this now. We can handle it on the main PR.

Reply:
Hm... I think it might be a good idea to change the parent from optional to mandatory in this context. Then we can add all this in the docstring.